traffic_monitor/datareq/crstate.go

```diff
@@ -84,10 +88,55 @@ func filterDirectlyPolledCaches(crstates tc.CRStates) tc.CRStates {
 	return filtered
 }
 
-func srvTRStateSelf(localStates peer.CRStatesThreadsafe, directlyPolledOnly bool) ([]byte, error) {
+func srvTRStateSelf(localStates peer.CRStatesThreadsafe, directlyPolledOnly bool, toData todata.TODataThreadsafe) ([]byte, error) {
 	if !directlyPolledOnly {
-		return tc.CRStatesMarshall(localStates.Get())
+		localStatesC := updateStatusAnycast(localStates, toData)
+		return tc.CRStatesMarshall(localStatesC)
 	}
-	unfiltered := localStates.Get()
+	unfiltered := updateStatusAnycast(localStates, toData)
 	return tc.CRStatesMarshall(filterDirectlyPolledCaches(unfiltered))
 }
+
+func updateStatusAnycast(localStates peer.CRStatesThreadsafe, toData todata.TODataThreadsafe) tc.CRStates {
+	localStatesC := localStates.Get()
+	toDataC := toData.Get()
+
+	for cache, _ := range localStatesC.Caches {
+		if _, ok := toDataC.ServerPartners[cache]; ok {
+			// all server partners must be available if they are in reported state
+			allAvailableV4 := true
+			allAvailableV6 := true
+			allIsAvailable := true
+			for partner, _ := range toDataC.ServerPartners[cache] {
+				if partnerState, ok := localStatesC.Caches[partner]; ok {
+					// a partner host is reported but is marked down for too high traffic or load
+					// this host also needs to be marked down to divert all traffic for their
+					// common anycast ip
+					if strings.Contains(partnerState.Status, "REPORTED") &&
+						strings.Contains(partnerState.Status, "too high") {
+						if !partnerState.Ipv4Available {
```

Comment:
This should check if `allAvailableV4` is still `true` and only set it to `false` if it is.